sources/packages/api/src/services/education-program-offering/education-program-offering-validation.service.ts

```diff
@@ -0,0 +1,131 @@
+      // Ensures that the object received is a class. This is needed to the
+      // proper validation metadata be available to the validation be performed.
+      const offeringModel = plainToClass(SaveOfferingModel, offering, {
+        enableImplicitConversion: true,
+      });
+      const validationsErrors = validateSync(offeringModel);
+      const allErrors: string[] = [];
+      const allWarnings: ValidationWarningResult[] = [];
+      validationsErrors.forEach((error) => {
+        const [errors, warnings] = this.extractErrorsAndWarnings(error);
+        allErrors.push(...errors);
+        allWarnings.push(...warnings);
+      });
+
+      const offeringStatus = this.getOfferingSavingStatus(
+        allErrors.length,
+        allWarnings.length,
+      );
+
+      if (!silently && !!allErrors.length) {
+        throw new CustomNamedError(
+          "The validated offerings have critical errors.",
+          OFFERING_VALIDATION_CRITICAL_ERROR,
+          allErrors,
+        );
+      }
+
+      return {
+        index,
+        offeringModel,
+        offeringStatus,
+        warnings: allWarnings,
+        errors: allErrors,
+      };
+    });
+  }
+
+  /**
+   * Inspect the validation error and its children checking when an error
+   * happen as has a warning context (must be considered a warning) or it is
+   * a critical error (has no warning context).
+   * @param error error to be inspected.
+   * @returns errors and warnings.
+   */
+  private extractErrorsAndWarnings(
+    error: ValidationError,
+  ): [errors: string[], warnings: ValidationWarningResult[]] {
+    const errors: string[] = [];
+    const warnings: ValidationWarningResult[] = [];
+    const flattenedErrors = flattenErrors(error);
+    flattenedErrors.forEach((error) => {
+      Object.keys(error.constraints).forEach((errorConstraintKey) => {
+        let associatedContext: ValidationWarning = undefined;
+        if (error.contexts) {
+          associatedContext = error.contexts[
+            errorConstraintKey
+          ] as ValidationWarning;
+        }
+        const errorDescription = error.constraints[errorConstraintKey];
+        if (associatedContext?.isWarning) {
+          warnings.push({
+            warningType: associatedContext.warningType,
+            warningMessage: errorDescription,
+          });
+        } else {
+          errors.push(errorDescription);
+        }
+      });
+    });
+    return [errors, warnings];
+  }
+
+  /**
+   * Defines the offering status based on the number of errors and warnings.
+   * @param totalErrors total critical errors.
+   * @param totalWarnings total warnings (offering can be saved but need Ministry review).
+   * @returns offering approved, pending or undefined case there is some critical error.
+   */
+  private getOfferingSavingStatus(
+    totalErrors: number,
+    totalWarnings: number,
+  ): OfferingStatus.Approved | OfferingStatus.CreationPending | undefined {
+    if (totalErrors > 0) {
```

Comment:
What is the significance of considering error here? if there is a validation error, then the offering object is never going to reach db. So we get a benefit of have one less if condition here.

Reply:
What should the method return if there is an error?
I believe that either at this point or the point where this method is called there will be an `if`, right?